Add support for IBM PowerAI DDL gpu allreduce

nvcastet opened the pull request:
* Include making FinalizerThread more generic using
  a queue of events to wait for.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/uber/horovod?pullRequest=323) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/uber/horovod?pullRequest=323) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/uber/horovod?pullRequest=323) it.</sub>

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -843,22 +936,37 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
         }
+
+#if HOROVOD_GPU_ALLREDUCE != 'D'
         if (timeline.Initialized()) {
```
Why do you side-step check for `timeline.Initialized()`?

nvcastet replied:
The DDL allreduce is a synchronous operation and uses its own Cuda streams. So we need to record an event on Horovod stream and wait for it before calling DDL to make all the work is done up to this point on the Horovod stream.
That is also why i need to start the finalizer_thread before all the events get added to the Horovod stream otherwise your timeline would be messed up because ddl is syncing the operation in the middle of all the events. Does that make sense?

alsrgv replied:
I see - makes sense.  I actually do some of the things along the similar lines in #329.  I think it makes sense to merge in refactorings from #329 first, and then your code change will actually be much simpler.

nvcastet replied:
ok will wait on #329.

nvcastet replied:
@alsrgv How is your PR #329 going?

alsrgv replied:
Waiting to hear back from a user in #312 to determine whether it works and whether it's faster.

alsrgv replied:
@nvcastet, the #329 has been pushed.  Please rebase/simplify this PR.

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -873,81 +981,44 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
-          ACTIVITY_END_ALL(entries, timeline)
-          RELEASE_EVENT(entries, queue_end_event);
-        }
+      RECORD_EVENT("ALLREDUCE_DONE", entries, done_event, stream, true)
```
Can we do this without changing the set of events we're tracking?

nvcastet replied:
This event is the done_event that was there before. You mean without adding an extra activity to the timeline? The bool set to true at the end of the call means it is the final event. We could not record the event as an activity in the finalizer thread. Would you prefer that?

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -1480,7 +1551,11 @@ void BackgroundThreadLoop(HorovodGlobalState& state) {
+#if HAVE_DDL
+  ddl_finalize();
+#else
   MPI_Finalize();
```
Does `ddl_finalize()` do `MPI_Finalize()` as well?

nvcastet replied:
Yes and also cleanup some MPI communicators so we cannot call MPI_Finalize before.

alsrgv commented on setup.py:
```diff
@@ -313,15 +313,29 @@ def get_nccl_dirs(build_ext, cuda_include_dirs, cuda_lib_dirs, cpp_flags):
 
+def get_ddl_dirs():
+    # Default DDL home
+    ddl_home = '/opt/DL/ddl'
```
We should make this location configurable.

nvcastet replied:
Yes we could. But It will most likely never change because our rpm and deb packages assume it gets setup to this location.

alsrgv replied:
Is it always rpm and deb, or you also have tarball with the libraries?

nvcastet replied:
Yes always rpm or dep.

alsrgv replied:
OK, it's fine to keep like that then.

nvcastet commented on horovod/common/operations.cc:
```diff
@@ -515,6 +528,18 @@ ncclDataType_t GetNCCLDataType(const std::shared_ptr<Tensor> tensor) {
+  case HOROVOD_FLOAT32:
+    return DDL_TYPE_FLOAT;
+  default:
+    throw std::logic_error("Type " + MPIDataType_Name(tensor->dtype()) +
```
DDL supports half-precision (16 bits) too. I didn't find a HOROVOD_FLOAT16 in this file. Is there a way to integrate half-precision with Horovod?

alsrgv replied:
There is support coming in `fp16` branch that is being evaluated right now.  If DDL support is there, it will be easy to extend to DDL then!

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -863,6 +896,18 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
 
         ACTIVITY_END_ALL(entries, timeline)
       }
+#else
```
Let's make it `elif GPU_ALLREDUCE == 'D'` for future extensibility.

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -562,8 +581,20 @@ ncclDataType_t GetNCCLDataType(const std::shared_ptr<Tensor> tensor) {
+  {                                                                            \
+    auto ddl_result = (op);                                                    \
+    if (ddl_result != DDL_SUCCESS) {                                           \
+      for (auto it = entries.begin(); it != entries.end(); it++) {             \
```
Can you change it to iterator, like the rest of new `*_CHECK()`

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -1342,7 +1395,11 @@ void BackgroundThreadLoop(HorovodGlobalState& state) {
+#if HAVE_DDL
+  ddl_finalize();
+#else
+   MPI_Finalize();
```
Can you add a comment that `ddl_finalize()` performs MPI finalization as well?

Nit: remove one extra space before `MPI`

nvcastet commented:
@alsrgv Could you look at my last comment in #329? I will wait for the fix to merge if one is needed.

nvcastet commented:
@alsrgv We have not yet run Horovod + DDL on a lot of nodes.
For up to 4 nodes here is the images/sec for HPM resnet101:
     GPUs: 1               2              4               8              16         
+NCCL2: 213.84	427.53	843.80	1665.70	3210.00
+DDL:      214.37	425.34	847.75	1680.50	3257.20
We think the DDL version will start getting better as we scale the number of nodes.

alsrgv commented:
@nvcastet, the fix has been landed.  Thanks for letting us know about the issue!

nvcastet commented:
@alsrgv No problem. I updated this PR accordingly and added a copy from input to output buffer for this case because DDL only has in-place allreduce now.

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -865,6 +898,18 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
+                                 "to be set to use DDL.");
+        }
+        DDL_CHECK(entries, "ddl_init", ddl_init(ddl_options))
+        horovod_global.ddl_initialized = true;
```
One more question.  Does this pass `test_horovod_allreduce_multi_gpu()`?  In the case of NCCL, we dynamically create a communicator based on the set of GPUs that will participate in the operation.  How does DDL handle that?

nvcastet replied:
It probably won't pass this test. Right now, DDL handles only one device per process.

nvcastet replied:
@alsrgv I am adding a check for the "test_horovod_allreduce_multi_gpu" to fail gracefully. Just curious, what is the use case for Horovod to support several devices per process?

nvcastet replied:
I pushed the check.

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -865,6 +899,21 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
+        horovod_global.ddl_initialized = true;
+        horovod_global.ddl_local_device_id = first_entry.device;
+      } else if (horovod_global.ddl_local_device_id != first_entry.device) {
+        throw std::logic_error("DDL does not support more than one GPU device per process.");
```
Instead of exception (on a background thread), can this do something like what DDL_CHECK does:
```
for (auto& e : (entries)) {
        timeline.End(e.tensor_name, nullptr);                                  \
        e.callback(Status::UnknownError(std::string(op_name) + " failed."));   \
}
return
```

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -865,6 +899,21 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
+        // Initialize DDL
+        auto ddl_options = std::getenv("DDL_OPTIONS");
+        if (ddl_options == nullptr) {
+          throw std::logic_error("DDL_OPTIONS env variable needs "
```
Similar to new the other comment, can this use mechanism similar to `DDL_CHECK`?

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -865,6 +899,21 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
+#elif HOROVOD_GPU_ALLREDUCE == 'D'
+      if (!horovod_global.ddl_initialized) {
+        // Initialize DDL
+        auto ddl_options = std::getenv("DDL_OPTIONS");
```
Can you give an example of DDL_OPTIONS?  Are they system-wide?

nvcastet replied:
They are setup by ddlrun (or you can set it up yourself too). I would refer to the DDL documentation to get the specific about DDL_OPTIONS.

alsrgv commented on horovod/common/operations.cc:
```diff
@@ -525,6 +533,18 @@ ncclDataType_t GetNCCLDataType(const std::shared_ptr<Tensor> tensor) {
+  case HOROVOD_FLOAT32:
+    return DDL_TYPE_FLOAT;
+  default:
+    throw std::logic_error("Type " + MPIDataType_Name(tensor->dtype()) +
```
`std::logic_error()` here is not desirable.

`GetMPIDataType()` and `GetNCCLDataType()` throw `std::logic_error()` because it indicates Horovod developer bug - they both cover all possible types allowed for GPU allreduce & other ops by TF and PyTorch, and if a user tries to use unsupported type he'll get a nice error before that in the stack.

Since here DDL is more restrictive, we should handle it nicer and similarly give a user error via a mechanism similar to `DDL_CHECK()`.

nvcastet replied:
I can do that

nvcastet commented:
@alsrgv I don't know if you see my comment of yesterday: "Just curious, what is the use case for Horovod to support several devices per process?"

alsrgv commented:
@nvcastet, model parallelism. The user can manually pin tensors on multiple GPUs and Horovod will make sure it will create an NCCL ring across the right set of GPUs for each op.
